### PR TITLE
sam4l: fix usart sleep during panics

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

Josh found a nice bug here. We special-cased panic for the usart manager,
but not quite enough. Specifically, panic doesn't use the usart peripheral
as the driver expects. When panic begins, it turns on the clock and
unconditionally expects the clock to stay on. If there's any outstanding
prints from debug, the debug `flush` will use the original usart driver to
complete these (critically, *after* the panic usart object has been created).
Thus when the `flush` print completes, the original driver (correctly) sees
that it has no more to do and disables usart clock out from underneath the
panic usart object.

The fix here is to just promote `IS_PANICING` to global state, and not tie
that state to a specific instance of usart manager.

### Testing Strategy

Panic'ing on Signpost modules

### Documentation Updated

- [x] ~~Kernel: Updated the relevant files in `/docs`, or no updates are required.~~
- [x] ~~Userland: Added/updated the application README, if needed.~~

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
Little concerning to see an undocumented global variable that requires `unsafe` be merged with no `P-` tag.

Can we not just use the `debug` module code for `panic` now? I don't think we _need_ the whole setup of the uart in io.rs. Are we concerned that someone would remove the console from the board file but still want panic to printf? That seems a little odd to me, given that our current panic is really only useful during debugging, when someone would likely want printf to work normally as well.

File: chips/sam4l/src/usart.rs [hunk 1]
```diff
@@ -281,11 +281,12 @@ pub struct USARTRegManager<'a> {
     clock: pm::Clock,
     rx_dma: Option<&'static dma::DMAChannel>,
     tx_dma: Option<&'static dma::DMAChannel>,
-    is_panic: bool,
 }
 
+static mut IS_PANICING: bool = false;
```
Comment (alevy) [inline chips/sam4l/src/usart.rs]: I agree with @bradjc. This should at least have been something like an `AtomicBool` so it wouldn't require `unsafe` to use.

**Comment by brghena** [comment]
Lets address the procedural failing that led to this. Clearly Pat and I saw it as fine, whereas you two see it as problematic.

For background, this was fixing a problem that Josh ran into when updating Signpost to the head of Tock and trying to get to low power. Pat created the PR and tested the change. I read over the code and decided it was fine. We then merged it so Josh could move forward. All of this happened over well less than an hour.

Should this have been P-Significant? Is the use of unsafe (or global variables) enough to push it to be a significant change?

If not, should P-Upkeep PRs have to wait a minimum amount of time before merging?

Is it fine if P-Upkeep changes sometimes require a second fix (such as `AtomicBool`) after merging? (via a second PR) Or should we try to make sure that comments like that get integrated into the original fix?

**Comment by ppannuto** [comment]
> Is it fine if P-Upkeep changes sometimes require a second fix (such as AtomicBool) after merging? (via a second PR) Or should we try to make sure that comments like that get integrated into the original fix?

I lean towards this answer. There's a reason that software has releases and a master branch. Also why there's generally a feature freeze / more caution leading up to a release.

I think this is a good example of something that addressed an immediate bug that was impeding progress that further refinement can now make better.

**Comment by alevy** [comment]
I lean towards `unsafe` in the kernel crate maybe pushing something over the edge to P-Significant, but I also agree that's it's not biggest deal in the world and we caught it and it's now being fixed.

It's fine for The Process(tm) to fail sometimes, and it's also fine for the process to work well and for it to still happen that bugs or other kinds of mistakes creep in. Let's just use this as a data point for what we all consider requiring more or less rigorous scrutiny as a PR.

**Pull request merged**